src/AppInstallerCommonCore/Yaml.cpp

```diff
@@ -0,0 +1,478 @@
+    {
+        Require(Type::Sequence);
+        return m_sequence.value()[index];
+    }
+
+    const Node& Node::operator[](size_t index) const
+    {
+        Require(Type::Sequence);
+        return m_sequence.value()[index];
+    }
+
+    size_t Node::size() const
+    {
+        switch (m_type)
+        {
+        case Type::Invalid:
+        case Type::None:
+        case Type::Scalar:
+            return 0;
+        case Type::Sequence:
+            return m_sequence->size();
+        case Type::Mapping:
+            return m_mapping->size();
+        }
+
+        THROW_HR(E_UNEXPECTED);
+    }
+
+    const std::vector<Node>& Node::Sequence() const
+    {
+        Require(Type::Sequence);
+        return m_sequence.value();
+    }
+
+    const std::multimap<Node, Node>& Node::Mapping() const
+    {
+        Require(Type::Mapping);
+        return m_mapping.value();
+    }
+
+    void Node::Require(Type type) const
+    {
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_YAML_INVALID_OPERATION, m_type != type);
+    }
+
+    void Node::Require(Type type1, Type type2) const
+    {
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_YAML_INVALID_OPERATION, m_type != type1 && m_type != type2);
+    }
+
+    std::string Node::as_dispatch(std::string*) const
+    {
+        return m_scalar;
+    }
+
+    int64_t Node::as_dispatch(int64_t*) const
+    {
+        return std::stoll(m_scalar);
+    }
+
+    bool Node::as_dispatch(bool*) const
+    {
+        if (m_scalar == "true")
+        {
+            return true;
+        }
+        else if (m_scalar == "false")
+        {
+            return false;
+        }
+        else
+        {
+            THROW_HR(APPINSTALLER_CLI_ERROR_YAML_INVALID_DATA);
+        }
+    }
+
+    Node Load(std::string_view input)
+    {
+        Wrapper::Parser parser(input);
+        Wrapper::Document document = parser.Load();
+
+        if (document.HasRoot())
+        {
+            return document.GetRoot();
+        }
+        else
+        {
+            return {};
+        }
+    }
+
+    Node Load(const std::string& input)
+    {
+        return Load(static_cast<std::string_view>(input));
+    }
+
+    Node Load(std::istream& input)
+    {
+        Wrapper::Parser parser(input);
+        Wrapper::Document document = parser.Load();
+
+        if (document.HasRoot())
+        {
+            return document.GetRoot();
+        }
+        else
+        {
+            return {};
+        }
+    }
+
+    Node Load(const std::filesystem::path& input)
+    {
+        std::ifstream stream(input, std::ios_base::in | std::ios_base::binary);
+        THROW_LAST_ERROR_IF(stream.fail());
+        return Load(stream);
+    }
+
+    Emitter::Emitter() :
+        m_document(std::make_unique<Wrapper::Document>(true))
+    {
+        SetAllowedInputs<InputType::BeginMap, InputType::BeginSeq>();
+    }
+
+    Emitter::Emitter(Emitter&&) = default;
+    Emitter& Emitter::operator=(Emitter&&) = default;
+
+    Emitter::~Emitter() = default;

```

Comment:
curious why not put these in the header?

Reply:
I'm using pImpl idioms here to remove the need for the libyaml header to be included in all of the consumers of our Yaml.h.  In order to do that, the std::unique_ptr<Emitter>'s destructor (and anything that requires the complete type) must only be referenced within the cpp.  So these functions are actually "implemented" here, rather than allowing them to be written out in every other TU and be folded by the compiler (or however they actually implement inline header functions).

---
In reply to: [468997684](https://github.com/Microsoft/pkg/pull/535#discussion_r468997684) [](ancestors = 468997684)